### PR TITLE
💲[Native Checkout] RewardCardView states VM tests

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

This adds the unit tests for `RewardCardContainerViewModel` and `RewardCardViewModel` for the work done in #779. This is just a separate PR to try to make it easier to review, it's still a lot though 😓 

# 🤔 Why

Tests gotta test.

# 🛠 How

- Added a `ButtonStyleType` which @ifbarrera and I chatted about doing so that we wouldn't be returning an actual styling function from the VM. Swift also doesn't support equality checks on function types so this is necessary for unit test assertions. We can decide where to put this type.
- Replicated the snapshot tests from #779 as unit tests in `RewardCardContainerViewModelTests` given that there's such a large amount of state combinations to cover. This is a slight deviation from how we usually write tests but I think I've put all of the necessary assertions in place for it to make sense.
- Tweaked some outputs slightly to make this work 👌 

# ✅ Acceptance criteria

- [ ] Tests pass.
- [ ] Tests make sense.
- [ ] I've covered all of the variations.

### CODE REVIEW

**Comment by justinswart** [comment]
[Reward States](https://trello.com/c/IUvuucUU/1448-reward-states)

File: Library/ViewModels/RewardCardViewModelTests.swift [hunk 1]
```diff
@@ -90,6 +91,31 @@ final class RewardCardViewModelTests: TestCase {
...
+        .template
+          |> Backing.lens.reward .~ reward
+          |> Backing.lens.rewardId .~ reward.id
+          |> Backing.lens.shippingAmount .~ 10
+          |> Backing.lens.amount .~ 700
+      )
+
+    self.vm.inputs.configureWith(
+      project: project,
+      rewardOrBacking: .left(reward)
```
Comment (ifbarrera) [inline Library/ViewModels/RewardCardViewModelTests.swift]: Just curious why wouldn't this be configured with the backing? 🤔

File: Kickstarter-iOS/Views/RewardCardContainerView.swift [hunk 2]
```diff
@@ -81,7 +81,9 @@ public final class RewardCardContainerView: UIView {
           .rewardCardView(self.rewardCardView, didTapWithRewardId: rewardId)
       }
 
-    self.pledgeButton.rac.title = self.viewModel.outputs.pledgeButtonTitleText
+    self.viewModel.outputs.pledgeButtonTitleText.observeValues { [weak self] text in
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Kickstarter-iOS/Views/RewardCardContainerView.swift]: Yeah, allows us to emit an optional from the VM.

File: Library/ViewModels/RewardCardContainerViewModelTests.swift [hunk 3]
```diff
@@ -9,57 +9,320 @@ import XCTest
...
+    self.pledgeButtonStyleType.assertValueCount(0)
+    self.pledgeButtonEnabled.assertValueCount(0)
+    self.pledgeButtonHidden.assertValueCount(0)
+    self.pledgeButtonTitleText.assertValueCount(0)
 
-    self.pledgeButtonEnabled.assertValues([true])
+    for (index, reward) in rewards.enumerated() {
+      let project = Project.cosmicSurgery
+        |> Project.lens.state .~ .successful
+        |> Project.lens.personalization.isBacking .~ false
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Library/ViewModels/RewardCardContainerViewModelTests.swift]: Good catch! This test is wrong, fill fix 👍

File: Library/ViewModels/RewardCardViewModelTests.swift (hunk 1 above)
Comment (justinswart) [reply Library/ViewModels/RewardCardViewModelTests.swift]: Just helped me reason about the tests better I guess, I'd like to remove the `Either` here at some point, not sure we need it.

**Pull request merged**